File: java/src/algorithms/sorting/quick_sort/QuickSort.java
package algorithms.sorting.quick_sort;

import algorithms.common.Cargo;
import java.util.ArrayList;

/**
 * Implementation of Quick Sort algorithm
 */
public class QuickSort {
    /**
     * Sorts an ArrayList of Cargo objects using Quick Sort
     * Time Complexity: O(n log n) average case, O(n^2) worst case
     * 
     * @param items ArrayList of Cargo objects to be sorted
     * @return time taken in nanoseconds
     */
    public static long sort(ArrayList<Cargo> items) {
        long start = System.nanoTime();

        ArrayList<Cargo> sorted = quickSort(items);
        // Copy sorted items back to original list
        for (int i = 0; i < items.size(); i++) {
            items.set(i, sorted.get(i));
        }

        long end = System.nanoTime();
        return end - start;
    }

    /**
     * Recursive quick sort implementation
     * 
     * @param items ArrayList to sort
     * @return sorted ArrayList
     */
    private static ArrayList<Cargo> quickSort(ArrayList<Cargo> items) {
        if (items.size() <= 1) {
            return items;
        }

        ArrayList<Cargo> lessThan = new ArrayList<>();
        ArrayList<Cargo> greaterThan = new ArrayList<>();

        // Use last element as pivot
        Cargo pivot = items.get(items.size() - 1);

        // Partition elements
        for (int i = 0; i < items.size() - 1; i++) {
            if (items.get(i).compareTo(pivot) < 0) {
                lessThan.add(items.get(i));
            } else {
                greaterThan.add(items.get(i));
            }
        }

        // Recursively sort partitions
        lessThan = quickSort(lessThan);
        greaterThan = quickSort(greaterThan);

        // Combine results
        ArrayList<Cargo> sorted = new ArrayList<>();
        sorted.addAll(lessThan);
        sorted.add(pivot);
        sorted.addAll(greaterThan);

        return sorted;
    }

    /**
     * Verify if the ArrayList is properly sorted
     * 
     * @param items ArrayList to check
     * @return true if sorted, false otherwise
     */
    public static boolean isSorted(ArrayList<Cargo> items) {
        for (int i = 1; i < items.size(); i++) {
            if (items.get(i - 1).compareTo(items.get(i)) > 0) {
                return false;
            }
        }
        return true;
    }
}

File: python/algorithms/sorting/quick_sort/quick_sort.py
"""
Quick Sort Algorithm Implementation
Author: Bjorn Melin
Date: 1/28/2025

Quick Sort is a divide-and-conquer algorithm that selects a pivot element 
and partitions the array into two sub-arrays, according to whether they 
are less than or greater than the pivot. The sub-arrays are then sorted 
recursively.

Time Complexity:
- Best Case: O(n log n)
- Average Case: O(n log n)
- Worst Case: O(n^2)

Space Complexity: O(log n)
"""

from typing import List

def quick_sort(arr: List[int]) -> List[int]:
    """
    Sorts an array of integers using the Quick Sort algorithm.

    Args:
        arr (List[int]): The list of integers to be sorted.

    Returns:
        List[int]: The sorted list of integers.
    """
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort(left) + middle + quick_sort(right)
